File: boosting/src/utils/MatrixUtils.java
package utils;

import java.util.Arrays;

public class MatrixUtils {
    private MatrixUtils() {
    }

    public static double multiply(double[] a, double[] b) {
        if (a.length != b.length) {
            throw new RuntimeException();
        }
        double res = 0;
        for (int i = 0; i < a.length; i++) {
            res += a[i] * b[i];
        }
        return res;
    }

    public static double[] multiply(double[] a, double coeff) {
        double[] res = Arrays.copyOf(a, a.length);
        for (int i = 0; i < a.length; i++) {
            res[i] = a[i] * coeff;
        }
        return res;
    }

    public static double[] subtract(double[] a, double[] b) {
        if (a.length != b.length) {
            throw new RuntimeException();
        }
        double[] res = new double[a.length];
        for (int i = 0; i < a.length; i++) {
            res[i] = a[i] - b[i];
        }
        return res;
    }

    public static double[] add(double[] a, double[] b) {
        if (a.length != b.length) {
            throw new RuntimeException();
        }
        double[] res = new double[a.length];
        for (int i = 0; i < a.length; i++) {
            res[i] = a[i] + b[i];
        }
        return res;
    }
}

File: boosting/src/utils/MatrixUtils.kt
package utils

import java.util.Arrays

object MatrixUtils {

    fun multiply(a: DoubleArray, b: DoubleArray): Double {
        if (a.size != b.size) {
            throw RuntimeException()
        }
        var res = 0.0
        for (i in a.indices) {
            res += a[i] * b[i]
        }
        return res
    }

    fun multiply(a: DoubleArray, coeff: Double): DoubleArray {
        val res = Arrays.copyOf(a, a.size)
        for (i in a.indices) {
            res[i] = a[i] * coeff
        }
        return res
    }

    fun subtract(a: DoubleArray, b: DoubleArray): DoubleArray {
        if (a.size != b.size) {
            throw RuntimeException()
        }
        val res = DoubleArray(a.size)
        for (i in a.indices) {
            res[i] = a[i] - b[i]
        }
        return res
    }

    fun add(a: DoubleArray, b: DoubleArray): DoubleArray {
        if (a.size != b.size) {
            throw RuntimeException()
        }
        val res = DoubleArray(a.size)
        for (i in a.indices) {
            res[i] = a[i] + b[i]
        }
        return res
    }
}
